course_UvA-DL/activation-functions/Activation_Functions.py

```diff
@@ -0,0 +1,824 @@
+# The third, most recently proposed activation function is Swish, which is actually the result of a large experiment with the purpose of finding the "optimal" activation function.
+# Compared to the other activation functions, Swish is both smooth and non-monotonic (i.e. contains a change of sign in the gradient).
+# This has been shown to prevent dead neurons as in standard ReLU activation, especially for deep networks.
+# If interested, a more detailed discussion of the benefits of Swish can be found in [this paper](https://arxiv.org/abs/1710.05941) [1].
+#
+# Let's implement the four activation functions below:
+
+# %%
+##############################
+
+
+class ReLU(ActivationFunction):
+
+    def forward(self, x):
+        return x * (x > 0).float()
+
+
+##############################
+
+
+class LeakyReLU(ActivationFunction):
+
+    def __init__(self, alpha=0.1):
+        super().__init__()
+        self.config["alpha"] = alpha
+
+    def forward(self, x):
+        return torch.where(x > 0, x, self.config["alpha"] * x)
+
+
+##############################
+
+
+class ELU(ActivationFunction):
+
+    def forward(self, x):
+        return torch.where(x > 0, x, torch.exp(x) - 1)
+
+
+##############################
+
+
+class Swish(ActivationFunction):
+
+    def forward(self, x):
+        return x * torch.sigmoid(x)
+
+
+##############################
+
+# %% [markdown]
+# For later usage, we summarize all our activation functions in a dictionary mapping the name to the class object.
+# In case you implement a new activation function by yourself, add it here to include it in future comparisons as well:
+
+# %%
+act_fn_by_name = {
+    "sigmoid": Sigmoid,
+    "tanh": Tanh,
+    "relu": ReLU,
+    "leakyrelu": LeakyReLU,
+    "elu": ELU,
+    "swish": Swish
+}
+
+# %% [markdown]
+# ### Visualizing activation functions
+#
+# To get an idea of what each activation function actually does, we will visualize them in the following.
+# Next to the actual activation value, the gradient of the function is an important aspect as it is crucial for optimizing the neural network.
+# PyTorch allows us to compute the gradients simply by calling the `backward` function:
+
+
+# %%
+def get_grads(act_fn, x):
+    """Computes the gradients of an activation function at specified positions.
+
+    Inputs:
+        act_fn - An object of the class "ActivationFunction" with an implemented forward pass.
+        x - 1D input tensor.
+    Output:
+        A tensor with the same size of x containing the gradients of act_fn at x.
+    """
+    x = x.clone().requires_grad_()  # Mark the input as tensor for which we want to store gradients
+    out = act_fn(x)
+    out.sum().backward()  # Summing results in an equal gradient flow to each element in x
+    return x.grad  # Accessing the gradients of x by "x.grad"
+
+
+# %% [markdown]
+# Now we can visualize all our activation functions including their gradients:
+
+
+# %%
+def vis_act_fn(act_fn, ax, x):
```

Comment:
thinking would be nice to define this sooner and after each Activation class show its chart
but the batch (all) plot below is also fine :]

Reply:
Yes, when creating the notebook we considered two versions: showing the plot of each activation one by one directly after its definition, or as a batch to give an overview. We decided for the latter version since it gives a better visualization to directly compare all activation functions with each other, and see their differences. I can see benefits for both versions, so if you have a strong preference, we can also change it.